### ISSUE TITLE
Pylance does not resolve list access before type check

### ISSUE THREAD

**Issue by Haini** [microsoft/pylance-release#648]
**tl;dr**
Accessing a returned list right at the function call gets not resolved correctly by pylance when using TypeVars as return types. 
```
var1 : type1 = fun()[0]
```
While this works:
```
var1 : List[type1] = fun()
var1[0].getPropertyX()
```
- VS Code Version: 1.51.1
- Python Plugin Version: 2020.11.371526539
- Pylance Version: 2020.11.2
- Python 3.8.5
---

I am not sure whether this is a bug or intended behavior, so I am posting this as a question: 

A method is called which returns a list of subclasses. The author already knows which subclass will be returned and types the variable that way. He also knows that only one element will be returned, so he already selects it right where he calls. This results in the following situation: 

**Actual behavior**
```
# I know that only >one< subclass SpecialFile will be returned, so I'd like to type selectedFiles
# that way already and select the first list element right where I call the function getSelectedFiles

# This doesn't work, pylance puts the following message:
# -- Expression of type "TFile" cannot be assigned to declared type "SpecialFile"
#-- "File" is incompatible with "SpecialFile"
seletedFilesNotWork : SpecialFile = OtherClass.getSelectedFiles("SpecialFile", fileListTest)[0]
```

**Correct behavior**
```
# I know that only subclass List[SpecialFile] will be returned, so I'd like to type selectedFiles
# that way already - this works
selectedFilesNew : List[SpecialFile] = OtherClass.getSelectedFiles("SpecialFile", fileListTest)

selectedFilesNew[0].specialFileFun()
selectedFilesNew[0].someSpecialMember
```

**Full minimal example** 
(could be more minimal I guess, [try it out on repl.it:](https://repl.it/join/cfcxqkcd-haini)):

```
from typing import List, TypeVar

TFile = TypeVar("TFile", bound="File")

#
#------------------------------------------------------------------------------
#
class File:
	def __init__(self, name : str):

		self.name : str = name

		self.selected : bool = True

	def isSelected(self):
		return self.selected
#
#------------------------------------------------------------------------------
#
class SpecialFile(File):

	def __init__(self, name : str):
		File.__init__(self, name)
		self.someSpecialMember = "TESTER"

	def specialFileFun(self):
		pass
#
#------------------------------------------------------------------------------
#
class CoolFile(File):

	def __init__(self, name : str):
		File.__init__(self, name)
	
	def coolFileFun(self):
		pass

class OtherClass:
	@staticmethod
	def getSelectedFiles(filter : str, files : List[File]) -> List[TFile]:
		"""
		docstring
		"""
		fileList : List[File] = []
		for file in files:			
			if file.isSelected() and type(file).__name__:
				fileList.append(file)

		return fileList



fileListTest : List[File] = [SpecialFile("bla"), CoolFile("blubb"), SpecialFile("dada")]

# This works fine of course, but then I am missing out on the typeHints
selectedFiles : List[File] = OtherClass.getSelectedFiles("SpecialFile", fileListTest)

# I know that only subclass SpecialFile will be returned, so I'd like to type selectedFiles
# already that way - this also works
selectedFilesNew : List[SpecialFile] = OtherClass.getSelectedFiles("SpecialFile", fileListTest)

selectedFilesNew[0].specialFileFun()
selectedFilesNew[0].someSpecialMember

# This doesn't work with pylance in vscode, it says: 
# Expression of type "TFile" cannot be assigned to declared type "SpecialFile"
#"File" is incompatible with "SpecialFile"
seletedFilesNotWork : SpecialFile = OtherClass.getSelectedFiles("SpecialFile", fileListTest)[0]

for file in selectedFilesNew:
    print(file.name)
```

**Comment by erictraut** [comment]
This sample is using type variables incorrectly. The `getSelectedFiles` is using a TypeVar `TFile` that is used only once in the method's signature, so there's no way for the type checker to "solve for" this type variable when the method is called. 

Pylance should be generating an error when you attempt to return `fileList` within this method's implementation because `List[File]` is not compatible with `List[TFile]`. The next release of Pylance will emit an error here.

There are two solutions I will recommend. The first is to update your `getSelectedFiles` to the following:
```python
@staticmethod
def getSelectedFiles(filter: Type[TFile], files: List[File]) -> List[TFile]:
    fileList: List[TFile] = []
    for file in files:
        if file.isSelected() and isinstance(file, filter):
            fileList.append(cast(TFile, file))

    return fileList
```

This is pretty clean except for the call to `cast`, which is required because the `isinstance` type narrowing doesn't apply type narrowing when the second argument is a type variable.

The second solution uses a new type-checking feature that has been proposed to the Python standards committee but has not yet been formally adopted. It's called a user-defined type guard. It's currently implemented only in Pylance (and Pyright) but not in other type checkers like mypy. If it becomes part of the standard, it will be implemented there as well. If it doesn't become part of the standard, its support will be removed from Pylance/Pyright. So use this at your own risk.

```python
if TYPE_CHECKING:
    # Import conditionally because typing_extensions doesn't
    # yet contain support for TypeGuard. It's defined only
    # in the type stub.
    from typing_extensions import TypeGuard


def is_file_type(file: File, filter: Type[TFile]) -> "TypeGuard[TFile]":
    return isinstance(file, filter)

class OtherClass:
    @staticmethod
    def getSelectedFiles(filter: Type[TFile], files: List[File]) -> List[TFile]:
        fileList: List[TFile] = []
        for file in files:
            if file.isSelected() and is_file_type(file, filter):
                fileList.append(file)

        return fileList
```

**Comment by Haini** [comment]
Thank you for the detailed explanation - this is very helpful!

I tried to go with your first suggestion and immediately receive two errors when changing the following lines

```
-		fileList : List[File] = []
+		fileList : List[TFile] = []
```

For using `TFile` instead of `File`
```
TypeVar "TFile" has no meaning in this context
```

For assigning the empty list to `fileList`:
```
Expression of type "List[File]" cannot be assigned to declared type "List[TFile]"
  TypeVar "_T" is invariant
    Type "File" cannot be assigned to type "TFile"
```

The error messages don't bring up any useful google results either, so I have to resort to this thread here again :-/

**Comment by erictraut** [comment]
I've recently fixed several bugs that would explain the errors you're seeing. These bugs were fixed a week ago in Pyright, but they are not in the latest version of Pylance yet. We didn't do a Pylance release this week because of the Thanksgiving holiday in the U.S. You'll see a new release next week that incorporates the latest bug fixes. Both of the solutions I suggested above work fine with the latest version.

**Comment by Haini** [comment]
Hi, that may very well be the case then! 

Happy Thanksgiving, I hope you can enjoy the holiday! And thanks again for your fast responses and help! :-)
